src/ZipkinBundle/Components/HttpClient/ContextRetriever.php

```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace ZipkinBundle\Components\HttpClient;
+
+use Zipkin\Propagation\TraceContext;
+
+/**
+ * ContextRetriever is an API to retrieve the context from the
+ * $options array (third parameter in the HttpClient request).
+ * Implementations can use different keys from the options or
+ * just decide to use the global context in the Scope API.
+ */
+interface ContextRetriever
```

Comment:
it this more a parser, or is it when something is actually having an object in a request attribute?

ex in brave Extractor is a parser (but it has to also return something when sampled only or baggage only)
secondly in instrumentation we often stash context into request attribute, but that is library specific

Reply:
So the usual use case in PHP SAPI (one process per request) is that we use the global context scope. In event loop we can't use globals, hence we need to pass somehow the context to the http client. Ideally it will be:

```php
$httpClient = new ZipkinHttpClient($httpClient, $tracing, InRequest())
...

$options = InRequest::wrapOptions($options, $context);
$httpClient->request('GET', 'http://zipkin.io', $options);
```

and internally the `wrapOptions` stash the context into the `$options['extra']` (as suggested by symfony on how to include custom data into a request) and this `ZipkinHttpClient` knows how to get the context from the `$options['extra']` array. It is very adhoc to Symfony.

If Symfony used a `RequestInterface` we could simply do:

```php
$httpClient = new ZipkinHttpClient($httpClient, $tracing, InRequest())
...
$httpClient->request(ZipkinRequest::withContext($request, $context));
```

Reply:
thanks for the notes. I suppose if some of these are default, maybe make a comment? ex above where options[extra] is used `// suggested by symfony on how to include custom data into a request`

similar for the rationale about RequestInterface but I was actually thinking about our own RequestInterface for parity in the callbacks. Originally in brave we used raw library types but that turned out not best.

Reply:
In the end I dropped the `ContextRetriever` for now as I don't seem to need it for now.